src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -382,39 +384,178 @@ See Zulip thread:
 https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/.E2.9C.94.20class_group.2Emk -/
 @[derive [inhabited, comm_ring]] def coordinate_ring : Type u := adjoin_root W.polynomial
 
+/-- The function field $R(W) := \mathrm{Frac}(R[W])$ of `W`. -/
+@[reducible] def function_field : Type u := fraction_ring W.coordinate_ring
+
+namespace coordinate_ring
+
+instance [subsingleton R] : subsingleton W.coordinate_ring :=
+(adjoin_root.mk_surjective W.monic_polynomial).subsingleton
+
 instance [is_domain R] [normalized_gcd_monoid R] : is_domain W.coordinate_ring :=
 (ideal.quotient.is_domain_iff_prime _).mpr $
 by simpa only [ideal.span_singleton_prime W.polynomial_ne_zero, ← gcd_monoid.irreducible_iff_prime]
    using W.irreducible_polynomial
 
-instance coordinate_ring.is_domain_of_field {F : Type u} [field F] (W : weierstrass_curve F) :
+instance is_domain_of_field {F : Type u} [field F] (W : weierstrass_curve F) :
   is_domain W.coordinate_ring :=
 by { classical, apply_instance }
 
-/-- The function field $R(W) := \mathrm{Frac}(R[W])$ of `W`. -/
-@[reducible] def function_field : Type u := fraction_ring W.coordinate_ring
-
 variables (x : R) (y : R[X])
```

Comment:
Hmm, ok, maybe mention this, even if the notation `R[X][X]` is already explained? In any case it is not so important.

Reply:
I'll add this to the docstring for `XY_ideal` that is defined downstream so it's all in one place.

Reply:
I think we should probably introduce local notations `R[Y] := R[X]` and `Y := polynomial.X` and use them systematically in these files. Or maybe `R[X][Y] := R[X][X]` and `Y := @polynomial.X (polynomial _) _` to ensure we never use `Y` to refer to the first variable. Definitely would look nicer in the paper :)

Reply:
I think this is a brilliant idea, but I have no idea how to make `notation` work with this, and `abbreviation` or `definition` just introduces another layer I have to unwrap.

Reply:
I made the change in [this branch](https://github.com/leanprover-community/mathlib/compare/associativity...associativity'?expand=1#diff-445ee9ab8837c7006ebbc11a3a8eaadcccdc0f204fdb3528aa8a85b826e7aadfR246-R252) and changed `X` to `Y` where appropriate. I used `Y := @polynomial.X (polynomial _) _` to ensure all `Y` refers to the second variable (and it caught a mistake), but I have to change it back to `polynomial.X` because Lean reports a "invalid notation: notation 'second_var' already declared" error whenever there are spaces between terms.

Reply:
Thanks, I've merged this into `associativity`. I completely forgot that you need backticks to declare notation.